libcfnet/stat_cache.c

```diff
@@ -349,3 +304,85 @@ const Stat *StatCacheLookup(const AgentConnection *conn, const char *file_name,
 
     return NULL;
 }
+
+/*********************************************************************/
+
+mode_t FileTypeToMode(const FileType type)
+{
+    /* TODO Match the order of the actual stat struct for easier mode */
+    int mode = 0;
+    switch (type)
+    {
+        case FILE_TYPE_REGULAR:
+            mode |= (mode_t) S_IFREG;
+            break;
+        case FILE_TYPE_DIR:
+            mode |= (mode_t) S_IFDIR;
+            break;
+        case FILE_TYPE_CHAR_:
+            mode |= (mode_t) S_IFCHR;
+            break;
+        case FILE_TYPE_FIFO:
+            mode |= (mode_t) S_IFIFO;
+            break;
+        case FILE_TYPE_SOCK:
+            mode |= (mode_t) S_IFSOCK;
+            break;
+        case FILE_TYPE_BLOCK:
+            mode |= (mode_t) S_IFBLK;
+            break;
+        case FILE_TYPE_LINK:
+            mode |= (mode_t) S_IFLNK;
+            break;
+    }
+
+    // mode is 0 if no file types matched
+    return mode;
+}
+
+/*********************************************************************/
+
+int StatParseResponse(const char *const buf, Stat *statbuf)
+{
+    assert(buf != NULL);
+    assert(statbuf != NULL);
+
+    // use intmax_t here to provide enough space for large values coming over the protocol
+    intmax_t d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12 = 0, d13 = 0;
+    int ret = sscanf(buf, "%1" PRIdMAX     // 01 statbuf->cf_type
+                          " %5" PRIdMAX    // 02 statbuf->cf_mode
+                          " %14" PRIdMAX   // 03 statbuf->cf_lmode
+                          " %14" PRIdMAX   // 04 statbuf->cf_uid
+                          " %14" PRIdMAX   // 05 statbuf->cf_gid
+                          " %18" PRIdMAX   // 06 statbuf->cf_size
+                          " %14" PRIdMAX   // 07 statbuf->cf_atime
+                          " %14" PRIdMAX   // 08 statbuf->cf_mtime
+                          " %14" PRIdMAX   // 09 statbuf->cf_ctime
+                          " %1" PRIdMAX    // 10 statbuf->cf_makeholes
+                          " %14" PRIdMAX   // 11 statbuf->cf_ino
+                          " %14" PRIdMAX   // 12 statbuf->cf_nlink
+                          " %18" PRIdMAX,  // 13 statbuf->cf_dev
+                     &d1, &d2, &d3, &d4, &d5, &d6, &d7,
+                     &d8, &d9, &d10, &d11, &d12, &d13);
+
+    if (ret < 13)
+    {
+        return -1;
+    }
+
+    statbuf->cf_type = (FileType) d1;
+    statbuf->cf_mode = (mode_t) d2;
+    statbuf->cf_lmode = (mode_t) d3;
+    statbuf->cf_uid = (uid_t) d4;
+    statbuf->cf_gid = (gid_t) d5;
+    statbuf->cf_size = (off_t) d6;
+    statbuf->cf_atime = (time_t) d7;
+    statbuf->cf_mtime = (time_t) d8;
+    statbuf->cf_ctime = (time_t) d9;
+    statbuf->cf_makeholes = (char) d10;
+    statbuf->cf_ino = d11;
+    statbuf->cf_nlink = d12;
+    statbuf->cf_dev = (dev_t)d13;
+
+    return 0;
```

Comment:
Should be `return ret`, right?

Reply:
I think `0` should be fine, since it returns `-1` in any other case.

Reply:
So there are only two options of what the function can return -- failure or success. Looks like a good candidate for `bool` return type to me.

Reply:
I agree : )